### PR TITLE
Topic recognition

### PR DESCRIPTION
**Pull request by WilliamBa29**
I have implemented a VQVAE model that has two convolutional neural network layers in the encoder and two convolutional neural network layers in the decoder. This model was employed on the OASIS brain data-set. The optimal model when comparing models on the validation data set (where the OASIS data-set comes with a predefined validation data set) was found to involve 10 embedding vectors, a latent space dimension of 10, a learning rate of 0.001 and a commitment cost of 0.3. The model achieves a mean SSIM on the test data-set of 0.9584.

### CODE REVIEW

**Comment by SiyuLiu0329** [comment]
- good documentation 
- reconstruction OK
- seems to be missing image generation?

**Comment by WilliamBa29** [comment]
@SiyuLiu0329 
I am a bit confused about your comment, 'seems to be missing image generation'. Does this refer to generating a manifold from the latent space as I was under the impression that we did not have to do this. I did show a reasonably clear image by reconstructing images from the test data. If you do want us to create manifold (or something else) do you want me to add this?

**Comment by SiyuLiu0329** [comment]
@WilliamBa29 
Yes your reconstruction is OK, and i saw the manifold images and they look quite nice. However, there is a generative aspect to VQVAEs (please see the end of https://keras.io/examples/generative/vq_vae/) which is also quite important. What you did with the manifold seems to be more inline with a traditional AE. Nevertheless, we will consider  the manifold bits for marking as well.

**Comment by WilliamBa29** [comment]
@SiyuLiu0329 

So do you want me to update the pull request with the generative part?

Do I also need to update the readme to mention this?

**Comment by WilliamBa29** [comment]
@SiyuLiu0329 

I have added functions to the files, ('dataset.py', 'modules.py', 'train.py', and 'predict.py') which extract the indexes of the embeddings, train a model to generate a prior, and progressively sample the prior (pixel by pixel) to generate new images. I have also updated the README file to reflect this. At the bottom of the README file you can see the generated images (they arent very good) resulting from the code but it is done.

**Comment by shakes76** [comment]
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate use and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem (generations empty) -2
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage 
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages, could be more informative -1
Progressive commits used
## Documentation
ReadMe acceptable but missing architecture diagrams -1
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
Feedback required, remove files that were added or changes outside your submission, there are some in algorithms and other changes to other students, please revert those -2
Request Description OK, could have more info -1

**Comment by WilliamBa29** [comment]
@shakes76 

I have now reverted the changes to the files by undoing the commit that led to it

**Comment by WilliamBa29** [comment]
I think there is some error with github. If you look at Files changed:   You will see that the changes made are the opposite of the changes made to files outside my submission under the commit, 'changed Modules to modules'.  'Files Changed' made now agree with the changes made under my reversal, which is the commit, "deleted changes to other files'.  Except, that before the reversal, I believe that under Files changed, it listed the changes made under the original commit, "changed Modules to modules' as the files changed for the pull request. In other words, it  seems that the original files dont want to agree with either the changed files under the commit, 'changed Modules to modules', or when I reverse these changes, which one would think is impossible. You can see by looking at my commits, that the only commit which changed files (i dont how this happened), prior to my reversal ( 'deleted changes to other files') was 'Changed Modules to modules' and that everyone of these is reversed, by looking at 'deleted changes to other files'. So I dont understand what the issue is. I looked at (in addition to the reversal) downloading a copy of all the files in your repository again, to see if I can replace my local repository's files with the original files in your repository in the required places, but when I download there is no directory for 'ISICs_Unet' so I did not do this. 'ISICs_Unet' appears to have been derived from ISICs_UNet' and appears to be name changed in some sense so maybe this is causing an issue

**Comment by WilliamBa29** [comment]
In continuance of what I was saying before, there seems to be a conflict between the ISICs_Unet in the recognition folder and ISICs_UNet under topic recognition branch (in your repo) they appear to be the same with slight changes. And also, it appears that one of these appears to looks similar to the case resulting from the changes 'changed Modules to modules' and the other one looks appears to resemble what would be the case before the changes in 'changed Modules to modules' (similarly it looks like what would result from the reversal, 'deleted changes to other files'. So there is some conflict here between the two somehow

**Comment by WilliamBa29** [comment]
I also just deleted an unnecessary test file from my submission. See the comments above to see how I attempted to solve the problem with other files being changed

**Comment by shakes76** [comment]
Unfortunately that deletes previous student files, I think all you needed was revert. Regardless, feedback marks will be awarded though, but I can't merge until it is fixed properly.
